[JIT] Enable ModuleDict non-literal indexing

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45716 [JIT] Enable ModuleDict non-literal indexing**

**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the left-hand side of the assignment has an interface type
annotation and if the dictionary contains only modules that implement
that interface, e.g. `module: ModuleInterface = self.module_dict[key]`.

The approach taken by this commit is that the annotation is passed
into `SugaredValue::getitem` as a type hint whose presence is used to
emit a special `prim::ModuleDictIndex` operator and set the output type
of the node appropriately.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

Commit:
[WIP] Enable ModuleDict non-literal indexing
Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Commit:
Update on "[WIP] Enable ModuleDict non-literal indexing"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

Commit:
Update on "[WIP] Enable ModuleDict non-literal indexing"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the `ModuleDict` is annotated with `Dict[str, X]`, where `X` is
a module interface type. These annotations must be expressed using a
class attribute named `__annotations__`, which is a `Dict[str, Type]`
where the keys are the names of module attributes and the values are
their types.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

Commit:
Update on "[WIP][JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the `ModuleDict` is annotated with `Dict[str, X]`, where `X` is
a module interface type. These annotations must be expressed using a
class attribute named `__annotations__`, which is a `Dict[str, Type]`
where the keys are the names of module attributes and the values are
their types.

The approach taken by this commit is that these annotations are stored
as "hints" along with the corresponding module attributes in the
`ConcreteSubmoduleTypeBuilder` instance for each module (which might be
a `ModuleDict`). These hints are passed into the `ModuleValue` that is
created for desugaring operations on submodules so that indexing into a
`ModuleDict` can be emitted as a getitem op into a dict emitted into the
graph that represents the `ModuleDict`.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

gmagogsfm commented:
Adding Lily as a shadow reviewer

SplitInfinity commented:
This solves the problem that motivated this PR, but I marked this as WIP because I'm not sure how to fit this nicely into the rest of TorchScript. It seems very narrowly targetted at the moment and I'm not sure it generalizes well.

In summary, this PR introduces the concept of a "hint" (probably an overloaded term but I welcome other suggestions) to `ConcreteModuleTypeBuilder` and its list of modules. The annotations present in `__annotations__` are added as hints for the corresponding attribute in the `ConcreteModuleTypeBuilder` instance of its parent; I think it makes more sense to view these hints/annotations as properties of the parent `Module` and not the submodule to which they apply. These hints are passed along to `ModuleValue` where they can be used to modify how certain ops are desugared. In this case of this PR, a `ModuleValue` that has such a hint emits a dictionary into the graph and emits `getitem` ops instead of statically desugaring attempts to index into the `ModuleDict`. I think this can easily be extended to support non-static indexing for `ModuleList`.

SplitInfinity commented on test/jit/test_module_containers.py:
```diff
@@ -428,3 +428,51 @@ def forward(self, inputs):
+        self.checkModule(m, (torch.randn(2, 2), "module"))
+
+
+        class ModDict(torch.nn.ModuleDict):
```
This seems like a weird use case but this is how Detectron2 intends to use this feature in order to avoid state_dict invalidation (the details of which I don't quite understand).

eellison commented on test/jit/test_module_containers.py:
```diff
@@ -428,3 +428,51 @@ def forward(self, inputs):
+
+
+        class Mod(torch.nn.Module):
+            __annotations__ = {"d": Dict[str, ModuleInterface]}
```
Thought: I wonder if we'd be better off with an annotation that isn't "Dict". It's not literally a dict. Maybe `Mapping[str, int]` (the built-in mypy type) ? Or just `nn.ModuleDict`. I don't think this is a blocker but something to think about if we were to make this public.

eellison commented on torch/csrc/jit/frontend/concrete_module_type.h:
```diff
@@ -74,14 +74,19 @@ class VISIBILITY_HIDDEN ConcreteModuleTypeBuilder {
+  void addModule(
+      std::string name,
+      std::shared_ptr<ConcreteModuleType> meta,
+      TypePtr hint = nullptr);
```
it's not clear to me from this api what `hint` is used for.

eellison replied:
bump

SplitInfinity replied:
Oops, forgot to rename, but I left a better explanation in a comment thread above. I'll add it in the comments in the code itself.

eellison commented on torch/csrc/jit/frontend/concrete_module_type.h:
```diff
@@ -74,14 +74,19 @@ class VISIBILITY_HIDDEN ConcreteModuleTypeBuilder {
   void addFailedAttribute(std::string name, std::string failureReason);
   void setIterableModuleKind(IterableModuleKind kind);
+  // Set a type hint on this Module (perhaps an InterfaceType).
+  void setHint(TypePtr hint);
```
Same here, this is a very generic api, i wouldn't know what it is used for if not for the context i'm reading in now. Could we make it more specific ? `setContainedModulesTypeHint` maybe ? i'm not sure

eellison commented on torch/csrc/jit/frontend/concrete_module_type.h:
```diff
@@ -123,13 +128,17 @@ class VISIBILITY_HIDDEN ConcreteModuleTypeBuilder {
 
     std::string name_;
     std::shared_ptr<ConcreteModuleType> meta_;
+    TypePtr hint_;
```
same comment applies here

eellison commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -217,6 +217,55 @@ std::shared_ptr<SugaredValue> PythonModuleValue::attr(
 
+ModuleValue::ModuleValue(
+    Value* self,
+    std::shared_ptr<ConcreteModuleType> concreteType,
```
Why do need `hint` if we already have `concreteType` ?

SplitInfinity replied:
I consider these contained type hints to be properties of the parent module (after all, that's where the annotation lives), so they are in the concrete type for the parent, and not in the submodule concrete type. They need to be passed into the `ModuleValue` representing the submodule. 

Given that we only allow a module to appear once in an entire module hierarchy, I guess it's fine to associate a contained type hint with the submodule itself, and not with its parent.

eellison replied:
I agree that we only want to store it with the parent module where the annotation lives; I think we don't need to propagate it to the child module. Once we've indexed into the parent module, the output becomes annotated interface type

SplitInfinity replied:
I don't think I understand. If the contained type hint is in the parent's concrete type, and `parent.sub_module_dict` is desugared to a `ModuleValue` object containing the concrete type of `sub_module_dict` (which does not have the contained type hint), how do we figure out whether to allow non-static indexing in `getitem` for the `sub_module_dict`?

eellison replied:
Okay so you have the dict:
```
def forward(self, x, key: str)
    return self[key].forward(x)
```

When we index into `self` we know its concrete module type, and then we set the output of `self[key]` to be the contained interface type.

SplitInfinity replied:
What about the other unit test in this PR:
```
            def forward(self, x: torch.Tensor, key: str) -> Any:
                return self.d[key].forward(x)
```
where `d` is a submodule that is a `ModuleDict` with a contained type hint annotated in the parent module?

eellison replied:
Yea, i think in this case `d` should have its contained Type Hint on its concrete module type

SplitInfinity replied:
But that concrete module type might be shared with other script modules instances to which the contained type hint does not apply.

eellison replied:
If the hint is part of the ConcreteModuleType itself, then it can't be shared with other module instances that it doesnt apply since we'll create a new type right ?

SplitInfinity replied:
```
    def get_or_create_concrete_type(self, nn_module):
        """
        Infer a ConcreteType from this `nn.Module` instance. Underlying JIT
        types are re-used if possible.
        """
        concrete_type_builder = infer_concrete_type_builder(nn_module)

        nn_module_type = type(nn_module)
        if nn_module_type not in self.type_store:
            self.type_store[nn_module_type] = []

        # Search the type store for an already-available JIT type
        known_types = self.type_store[nn_module_type]
        for known_type in known_types:
            if known_type.equals(concrete_type_builder):
                return known_type

        # We didn't find anything; generate a new JIT type from this concrete type
        concrete_type = concrete_type_builder.build()
        self.type_store[nn_module_type].append(concrete_type)
        return concrete_type
```

Ah, so you are saying that by setting the hint on some concrete type X, it will not be shared because 
```
        for known_type in known_types:
            if known_type.equals(concrete_type_builder):
                return known_type
```
will not find any type that matches?

SplitInfinity replied:
I think that makes sense.

eellison commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -252,6 +301,11 @@ SugaredValuePtr ModuleValue::getitem(
         }
       }
       throw ErrorReport(loc) << "Key Error, " << idx_str;
+    } else if (hint_) {
```
We already have `concreteType_`, can we just look it up there ?

SplitInfinity replied:
In the current implementation, these contained type hints live on the parent concrete type of the module to which they apply, or in the case of an annotation on `self`, in the concrete type of the module itself. In one case, the hint needs to be threaded through and in another case it doesn't. 

Given that a module can only appear once in a module hierarchy, storing these contained type hints in the concrete types of the modules to which they apply would also work and make the code cleaner, but might be more confusing to understand.

eellison commented on torch/csrc/jit/python/python_sugared_value.h:
```diff
@@ -206,6 +208,13 @@ struct VISIBILITY_HIDDEN ModuleValue : public SugaredValue {
+  TypePtr hint_;
+  // If this is a ModuleDict with a hint_, this is a pointer to a dictionary in
+  // the graph that can be reused across lookups.
+  std::shared_ptr<SimpleValue> dict_{nullptr};
```
I don't know if we can do this in the presence of writes to the moduledict. I'd be curious if CSE works well here (it might not)

SplitInfinity replied:
I can add a test for that and emit the dict for every lookup if needed.

SplitInfinity replied:
I tried writing to the `ModuleDict` in the body of `forward` using both `[]` and `ModuleDict.update` and neither compiled, so it seems we don't have to worry about this for now? I think it would be a good thing to add to futureproof this work but it's hard to test given that you can't actually update the `ModuleDict` in TorchScript.

eellison replied:
You can't index into it with `[]` but you can use setAttr. still it's extremely rare, i'm not sure what the best solution is here. One possible solution is to have add a new op that takes in a string and the ModuleDict

eellison replied:
More i'm thinking about i think we should probably just add a `prim::IndexModuleDict` op or something along those lines

SplitInfinity replied:
What would the schema for the op be? I'd like to create something like the generic list ops, but with one template parameter for the `ModuleDict`'s type (since it's not really a dict), and another for the return type:
```
prim::ModuleDictIndex(t mod_dict, str ind) -> u
```

eellison replied:
Yea, it would have to be a special op, like IndexTuple

eellison commented on torch/jit/_recursive.py:
```diff
@@ -147,13 +151,16 @@ def infer_type(name, item):
-            assert attr_type.is_interface_type()
-            # if the type can be inferred, it should be a module interface type
-            sub_concrete_type = torch._C.ConcreteModuleType.from_jit_type(attr_type)
+            if attr_type.is_interface_type():
```
What's going on here exactly ? i'm not following

SplitInfinity replied:
So as far as I know, at the moment, if a submodule is a key in `__annotations__`, it can only map to a module interface type. This PR changes that (it can map to a contained type hint also), so we need to handle both cases and can no longer be certain that it must be a module interface type.

eellison reviewed (commented):
Looks great! A few comments:
- it would be nice to get rid of all of the threading of `hint` when we already have the concrete module type
- IMO it would be good to make the API / variable name more specific. 

Also, I don't think we're doing any type checking right now that the contained modules actually are compatible with the annotated type.

Re: repeated dict constructions, I would be curious if CSE does a good job of cleaning up the repeated constructions. I wouldn't be surprised if it didn't, due to aliasing limitations. We should figure that out, and if not, consider introducing an op or looking to what it would take to fix the aliasing limitations.

SplitInfinity commented:
> Looks great! A few comments:
> 
> * it would be nice to get rid of all of the threading of `hint` when we already have the concrete module type
> * IMO it would be good to make the API / variable name more specific.

Agreed, I will update the PR to rename hint to something more specific.

> Also, I don't think we're doing any type checking right now that the contained modules actually are compatible with the annotated type.

I do recall running across a type checking error while working on this PR. I'll add a test with mismatched types to confirm and add type checking if there is none.

Commit:
Update on "[WIP][JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the `ModuleDict` is annotated with `Dict[str, X]`, where `X` is
a module interface type. These annotations must be expressed using a
class attribute named `__annotations__`, which is a `Dict[str, Type]`
where the keys are the names of module attributes and the values are
their types.

The approach taken by this commit is that these annotations are stored
as "hints" along with the corresponding module attributes in the
`ConcreteSubmoduleTypeBuilder` instance for each module (which might be
a `ModuleDict`). These hints are passed into the `ModuleValue` that is
created for desugaring operations on submodules so that indexing into a
`ModuleDict` can be emitted as a getitem op into a dict emitted into the
graph that represents the `ModuleDict`.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the `ModuleDict` is annotated with `Dict[str, X]`, where `X` is
a module interface type. These annotations must be expressed using a
class attribute named `__annotations__`, which is a `Dict[str, Type]`
where the keys are the names of module attributes and the values are
their types.

The approach taken by this commit is that these annotations are stored
as "hints" along with the corresponding module attributes in the
`ConcreteSubmoduleTypeBuilder` instance for each module (which might be
a `ModuleDict`). These hints are passed into the `ModuleValue` that is
created for desugaring operations on submodules so that indexing into a
`ModuleDict` can be emitted as a getitem op into a dict emitted into the
graph that represents the `ModuleDict`.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

SplitInfinity commented:
I just pushed an update with checks for whether the contained type annotation is followed. I looked into whether the dict needs to be emitted before every lookup, but it seems that a `ModuleDict` cannot be written to in TorchScript with `[]` or `.update` so I don't think we need to worry about this.

eellison commented on torch/csrc/jit/frontend/concrete_module_type.cpp:
```diff
@@ -269,9 +297,10 @@ void ConcreteModuleTypeBuilder::addBuiltinFunction(
-    std::shared_ptr<ConcreteModuleType> meta) {
-  modules_.emplace_back(
-      ConcreteModuleTypeBuilder::ModuleInfo{std::move(name), std::move(meta)});
+    std::shared_ptr<ConcreteModuleType> meta,
```
The api here is seems to be a little off from everything else, where we exclusively use the `set{...}` functions. Is there a reason for the divergence ?

SplitInfinity replied:
Yeah, so in accordance with considering the annotation as a property of the parent and not of the submodule to which it applies, the annotation is added here, in `addModule` rather than directly on the parent. We could add it later using something like `ConcreteModuleTypeBuilder ::setSubmoduleContainedTypeHint(name, hint)` but it seems like an unnecessary second function call. I don't feel too strongly about this though so I'd be happy to make this change.

eellison commented on torch/csrc/jit/frontend/concrete_module_type.cpp:
```diff
@@ -269,9 +297,10 @@ void ConcreteModuleTypeBuilder::addBuiltinFunction(
-  modules_.emplace_back(
-      ConcreteModuleTypeBuilder::ModuleInfo{std::move(name), std::move(meta)});
+    std::shared_ptr<ConcreteModuleType> meta,
+    TypePtr containedTypeHint) {
```
when would we be adding `containedTypeHint` only to one specific module ?

SplitInfinity replied:
```
class Mod(torch.nn.Module):
            __annotations__ = {"d": Dict[str, ModuleInterface]}

            def __init__(self):
                super().__init__()
                self.d = torch.nn.ModuleDict({"module": ImplementsInterface()})

            def forward(self, x: torch.Tensor, key: str) -> Any:
                return self.d[key].forward(x)
```

The way I've been thinking about it is that when building the concrete type for `Mod`, we add `Dict[str, ModuleInterface]` as a type hint to the `ConcreteModuleType` for `Mod` *for its attribute named `d`*. I haven't been thinking about the hint as something that applies to the module referred to by `d` in all circumstances.

When/if we eventually fix module aliasing, the module referred to by `d` could appear in multiple places, so setting the hint directly on its `ConcreteModuleType` wouldn't make sense.

eellison commented on torch/csrc/jit/frontend/concrete_module_type.cpp:
```diff
@@ -269,9 +297,10 @@ void ConcreteModuleTypeBuilder::addBuiltinFunction(
-  modules_.emplace_back(
-      ConcreteModuleTypeBuilder::ModuleInfo{std::move(name), std::move(meta)});
+    std::shared_ptr<ConcreteModuleType> meta,
+    TypePtr containedTypeHint) {
```
Theres `setContainedTypeHint` which sets the type hint for all contained modules, and now we have an argument `containedTypeHint` which is only for one specific module. Maybe rename `containedTypeHint` variable name to be more specific to the module? At first glance i thought the variable was being used for the same purposed as `setContainedTypeHint`, but it's not

SplitInfinity replied:
Hmm, yeah, there are two sort of overlapping concepts here:
1. `ConcreteModuleTypeBuilder::containedTypeHint_`: This is a type hint for all contained modules. The `ConcreteModuleTypeBuilder` should have its `iterableModuleKind_` set to `IterableModuleKind::DICT` (but I think we can handle `IterableModuleKind::LIST` easily as well) for this to have meaning. This is the type hint that gets set and used when `self` is annotated:
```
        # Test annotation of self.
        class ModDict(torch.nn.ModuleDict):
            # Calls setContainedType on the ConcreteModuleTypeBuilder associated with ModDict
            __annotations__ = {"self": Dict[str, ModuleInterface]}

            def __init__(self):
                super().__init__({"module": ImplementsInterface()})

            def forward(self, x: torch.Tensor, key: str) -> Any:
                return self[key].forward(x)
```
2. `ConcreteModuleTypeBuilder::ModuleInfo::containedTypeHint_`: This is a type hint that applies only to the submodule associated with the `ModuleInfo` instance only when it is accessed as a submodule of the module associated with the `ConcreteModuleTypeBuilder` instance. That's why the type hint is threaded through to `ModuleValue` instead of being accessed from the `ModuleValue::concreteType_`; it does not exist on the concrete type of the submodule.
```
        class Mod(torch.nn.Module):
            # Calls addModule on the ConcreteModuleTypeBuilder associated 
            # with ModDict and passes Dict[str, ModuleInterface] as the hint for the 
            # submodule named d
            __annotations__ = {"d": Dict[str, ModuleInterface]}

            def __init__(self):
                super().__init__()
                self.d = torch.nn.ModuleDict({"module": ImplementsInterface()})

            def forward(self, x: torch.Tensor, key: str) -> Any:
                return self.d[key].forward(x)
```

I am reluctant to unify them because when/if we fix module aliasing and allow the same script module to be referred to in multiple places, always setting the contained type hint on the concrete type of the module it refers to (option 1 above) will no longer make sense.

SplitInfinity replied:
Disregard this, I think type sharing is not a concern here. As you pointed out below, setting a type hint on an instance of `ConcreteModuleTypeBuilder` will cause `equals` to return `false` when `ConcreteTypeStore.get_or_create_concrete_type` tries to find an existing type that matches the concrete type for the module currently being compiled.

SplitInfinity replied:
Disregard this, I think you're right in saying that type sharing is not a concern here because setting a type hint on a concrete type will make it compare unequal to any concrete type that would be equal to it in the absence of that type hint.

eellison commented on torch/csrc/jit/frontend/concrete_module_type.h:
```diff
@@ -83,6 +86,8 @@ class VISIBILITY_HIDDEN ConcreteModuleTypeBuilder {
   void addFailedAttribute(std::string name, std::string failureReason);
   void addIgnoredAttribute(std::string name);
   void setIterableModuleKind(IterableModuleKind kind);
+  // Set a type hint on this Module (perhaps an InterfaceType).
```
mind clarifying the comment ? does this apply to the module itself or its contained modules

SplitInfinity replied:
Yep, I will include the explanations from these comment threads in the code itself.

eellison reviewed (commented):
Looks good, just want to clarify a few things to make sure i'm understanding

ppwwyyxx commented:
It seems extra type checks were added after last time I tried this PR. The submodules in the ModuleDict must have the same signature including their argument name.
For our use case we can guarantee that indexing will only be used on submodules that match the declared interface (in fact it's a constant list of submodules), but submodules with a different signature/argument name may appear in the ModuleDict. Is it possible to relax the check for such situation?

eellison commented:
@ppwwyyxx could you try wrapping the non-conforming modules in another module ? 

```
@torch.jit.interface
class ModuleInterface(torch.nn.Module):
    def forward(self, inp: Any) -> Any:
        pass

class ShouldConformButDoesnt(torch.nn.Module):
    def forward(self, bad_name: Any) -> Any:
        pass


class WrapperModule(torch.nn.Module):
    def __init__(self, mod):
        super(WrapperModule, self).__init__()
        self.mod = mod

    def forward(self, bad_name: Any) -> Any:
        return self.mod(bad_name)
```

Interfaces in general have a good number of issues. Ideally we'd do something like protocols instead https://mypy.readthedocs.io/en/stable/protocols.html but it's a lot of work and isnt actually needed that often. You can use https://github.com/pytorch/pytorch/pull/45645/  to do the conversion to the wrapper module.

ppwwyyxx commented:
Yeah mismatched names should be relatively easy to address but I'm more concerned with a case where the return types also mismatched (despite the indexing syntax will only be used on submodules that have correct interface)

SplitInfinity commented on torch/csrc/jit/frontend/concrete_module_type.cpp:
```diff
@@ -92,7 +92,14 @@ ConcreteModuleType::ConcreteModuleType(ConcreteModuleTypeBuilder data)
+  bool equals = lhs.name_ == rhs.name_;
+  equals &= lhs.meta_->equals(*rhs.meta_);
+
+  if (lhs.containedTypeHint_ && rhs.containedTypeHint_) {
```
Oops, yeah.

SplitInfinity commented:
> Yeah mismatched names should be relatively easy to address but I'm more concerned with a case where the return types also mismatched (despite the indexing syntax will only be used on submodules that have correct interface)

It's not ideal but what about using `Any` return type? We recently made an improvement to type refinement (#46062) that would allow you to refine `Any` to any TorchScript type and have the same code work in eager mode.

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the `ModuleDict` is annotated with `Dict[str, X]`, where `X` is
a module interface type. These annotations must be expressed using a
class attribute named `__annotations__`, which is a `Dict[str, Type]`
where the keys are the names of module attributes and the values are
their types.

The approach taken by this commit is that these annotations are stored
as "hints" along with the corresponding module attributes in the
`ConcreteSubmoduleTypeBuilder` instance for each module (which might be
a `ModuleDict`). These hints are passed into the `ModuleValue` that is
created for desugaring operations on submodules so that indexing into a
`ModuleDict` can be emitted as a getitem op into a dict emitted into the
graph that represents the `ModuleDict`.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the `ModuleDict` is annotated with `Dict[str, X]`, where `X` is
a module interface type. These annotations must be expressed using a
class attribute named `__annotations__`, which is a `Dict[str, Type]`
where the keys are the names of module attributes and the values are
their types.

The approach taken by this commit is that these annotations are stored
as "hints" along with the corresponding module attributes in the
`ConcreteSubmoduleTypeBuilder` instance for each module (which might be
a `ModuleDict`). These hints are passed into the `ModuleValue` that is
created for desugaring operations on submodules so that indexing into a
`ModuleDict` can be emitted as a getitem op into a dict emitted into the
graph that represents the `ModuleDict`.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

ppwwyyxx commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -217,6 +217,54 @@ std::shared_ptr<SugaredValue> PythonModuleValue::attr(
+    for (size_t i = 0; i < self_type->numAttributes(); ++i) {
+      const auto& attr_type = self_type->getAttribute(i);
+      if (attr_type->is_module()) {
+        if (!attr_type->isSubtypeOf(dict_type->getValueType())) {
+          auto loc = self->node()->sourceRange();
+          throw ErrorReport(loc)
+              << "Attribute " << self_type->getAttributeName(i)
+              << " is not of annotated type "
+              << dict_type->getValueType()->annotation_str();
```
Would it be worthwhile to print more detailed reasons using the 2nd argument of `isSubtypeOf`? I don't actually know what's in the message but if it can tell me which method/argument does not match it would be very helpful.

ppwwyyxx commented:
> It's not ideal but what about using Any return type?

Thanks! That works well! We don't always get to change the return type (e.g. if the module is defined in pytorch) but wrapping the module as @eellison suggests can solve that. We now can script our backbone (D24473591)

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the `ModuleDict` is annotated with `Dict[str, X]`, where `X` is
a module interface type. These annotations must be expressed using a
class attribute named `__annotations__`, which is a `Dict[str, Type]`
where the keys are the names of module attributes and the values are
their types.

The approach taken by this commit is that these annotations are stored
as "hints" along with the corresponding module attributes in the
`ConcreteSubmoduleTypeBuilder` instance for each module (which might be
a `ModuleDict`). These hints are passed into the `ModuleValue` that is
created for desugaring operations on submodules so that indexing into a
`ModuleDict` can be emitted as a getitem op into a dict emitted into the
graph that represents the `ModuleDict`.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the `ModuleDict` is annotated with `Dict[str, X]`, where `X` is
a module interface type. These annotations must be expressed using a
class attribute named `__annotations__`, which is a `Dict[str, Type]`
where the keys are the names of module attributes and the values are
their types.

The approach taken by this commit is that these annotations are stored
as "hints" along with the corresponding module attributes in the
`ConcreteSubmoduleTypeBuilder` instance for each module (which might be
a `ModuleDict`). These hints are passed into the `ModuleValue` that is
created for desugaring operations on submodules so that indexing into a
`ModuleDict` can be emitted as a getitem op into a dict emitted into the
graph that represents the `ModuleDict`.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

SplitInfinity commented:
I just pushed an update that unifies the two overlapping contained type hints that I had introduced upon realizing that you are right in saying that type sharing is not an issue.

eellison commented on torch/csrc/jit/frontend/concrete_module_type.h:
```diff
@@ -172,6 +173,13 @@ class VISIBILITY_HIDDEN ConcreteModuleTypeBuilder {
   py::object pyClass_;
 
+  // A type hint on this Module indicating the type of what it contains.
+  // This only makes sense on ModuleDict (where it might be Dict[str,
```
I suppose since we don't allow nn.Modules as specifyable types, it always has to be a InterfaceType

eellison commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -217,6 +217,54 @@ std::shared_ptr<SugaredValue> PythonModuleValue::attr(
+  auto containedTypeHint = concreteType_->getContainedTypeHint();
+
+  if (containedTypeHint) {
+    // For now, only dict type containedTypeHints are supported. Generate and
```
I think it might be better to defer generating the Dict until the value is actually indexed into. Right now, we're eagerly creating the dict so that each time the module value is referenced we create a new dict. It also locates the code for type-checking indexing more where you'd expect to be, instead of the generic ModuleValue creation

eellison commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -217,6 +217,54 @@ std::shared_ptr<SugaredValue> PythonModuleValue::attr(
+    std::vector<std::string> submodule_names;
+    const auto& self_type = concreteType_->getJitType()->expect<ClassType>();
+    for (size_t i = 0; i < self_type->numAttributes(); ++i) {
+      const auto& attr_type = self_type->getAttribute(i);
```
Can you refactor this to use `getSugaredDict` ? It will already generate all of the module keys, values for you

SplitInfinity replied:
I was able to remove this now that we're using a special op.

eellison commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -217,6 +217,54 @@ std::shared_ptr<SugaredValue> PythonModuleValue::attr(
+    for (const auto& name : submodule_names) {
+      auto name_v = insertConstant(*graph, name);
+      Value* module_v = graph->insertGetAttr(self_, name);
+      module_v->setType(dict_type->getValueType());
```
I don't think we should reset the type, I don't know if that is needed since we're only using it to construct the dictionary. I also dont know if it has any unintended consequences.

SplitInfinity replied:
I was able to remove this now that we're using a special op, but I did need to do something similar when compiling after deserialization. At that time, there are no `ConcreteTypeBuilder` and `ConcreteType` instances, and all modules are compiled as class types. Consequently, the output type of the indexing operation is taken from the `prim::ModuleDictIndex` operator I introduced (which is `Any`, not specific enough), which I reset to the correct interface type.

eellison reviewed (commented):
Looks great, a few more questions / comments

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the `ModuleDict` is annotated with `Dict[str, X]`, where `X` is
a module interface type. These annotations must be expressed using a
class attribute named `__annotations__`, which is a `Dict[str, Type]`
where the keys are the names of module attributes and the values are
their types.

The approach taken by this commit is that these annotations are stored
as "hints" along with the corresponding module attributes in the
`ConcreteSubmoduleTypeBuilder` instance for each module (which might be
a `ModuleDict`). These hints are passed into the `ModuleValue` that is
created for desugaring operations on submodules so that indexing into a
`ModuleDict` can be emitted as a getitem op into a dict emitted into the
graph that represents the `ModuleDict`.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

SplitInfinity commented:
I just uploaded a new version that uses a special operator `prim::ModuleDictIndex` instead of emitting the dict into the IR, but there were some additional consequences:
- I had to add a field for a contained type hint to `ClassType` so that it can be used when there is no `ConcreteType` instance available (e.g. when compiling something after deserialization)
- I had to modify `python_print.cpp` to emit this op as a regular subscript expression and emit the `__annotations__` entries for the contained type hints
- I had to modify `import_source.cpp` to parse `__annotations__` and add contained type hints back to reconstructed class types
- I had to modify `sugared_value.cpp` to add special desugaring logic in `SimpleValue` for emitting `prim::ModuleDictIndex` because it seems that even module types are treated like regular class types when compiled after deserialization, so the modifications I made to `ModuleValue` in `python_sugared_value.cpp` are not used.

I'm worried that these changes might be too intrusive. I'll request review after tests have passed.

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the `ModuleDict` is annotated with `Dict[str, X]`, where `X` is
a module interface type. These annotations must be expressed using a
class attribute named `__annotations__`, which is a `Dict[str, Type]`
where the keys are the names of module attributes and the values are
their types.

The approach taken by this commit is that these annotations are stored
as "hints" along with the corresponding module attributes in the
`ConcreteSubmoduleTypeBuilder` instance for each module (which might be
a `ModuleDict`). These hints are passed into the `ModuleValue` that is
created for desugaring operations on submodules so that indexing into a
`ModuleDict` can be emitted as a getitem op into a dict emitted into the
graph that represents the `ModuleDict`.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the left-hand side of the assignment has an interface type
annotation and if the dictionary contains only modules that implement
that interface, e.g. `module: ModuleInterface = self.module_dict[key]`.

The approach taken by this commit is that the annotation is passed
into `SugaredValue::getitem` as a type hint whose presence is used to
emit a special `prim::ModuleDictIndex` operator and set the output type
of the node appropriately.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

SplitInfinity commented:
Based on our offline discussion, I went a different route and uploaded a new iteration in which the type annotation must be inline with the assignment instead of at the class level for the entire `ModuleDict`:
```
value: ModuleInterface = self.module_dict[key]
```
`emitSugaredExpr` already has a `type_hint` parameter, so I threaded that through to `emitSubscript` and `SugaredValue::getitem` as an alternative to all of this contained type hint stuff. Now that we know the alternative 😨, I think it is easier to justify modifications to the `SugaredValue` interface. There are still some changes required in `python_print.cpp` to print out the type annotation and `prim::ModuleDictIndex` correctly but I was able to undo all of the other changes 🥳 .

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the left-hand side of the assignment has an interface type
annotation and if the dictionary contains only modules that implement
that interface, e.g. `module: ModuleInterface = self.module_dict[key]`.

The approach taken by this commit is that the annotation is passed
into `SugaredValue::getitem` as a type hint whose presence is used to
emit a special `prim::ModuleDictIndex` operator and set the output type
of the node appropriately.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the left-hand side of the assignment has an interface type
annotation and if the dictionary contains only modules that implement
that interface, e.g. `module: ModuleInterface = self.module_dict[key]`.

The approach taken by this commit is that the annotation is passed
into `SugaredValue::getitem` as a type hint whose presence is used to
emit a special `prim::ModuleDictIndex` operator and set the output type
of the node appropriately.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

eellison commented on torch/csrc/jit/frontend/sugared_value.cpp:
```diff
@@ -393,6 +394,14 @@ SugaredValuePtr SimpleValue::getitem(
         g.insert(aten::select, {val, 0, idx}, {}, loc));
   } else if (auto class_type = val_type->cast<ClassType>()) {
+    // Check if this is an indexing operation enabled by a type hint.
+    if (class_type->is_module() && type_hint) {
```
Maybe add a comment to the extent of, we’ve already type checked this upon initial emission

eellison commented on torch/csrc/jit/python/python_sugared_value.cpp:
```diff
@@ -252,6 +254,27 @@ SugaredValuePtr ModuleValue::getitem(
+      }
+
+      // Emit a prim::ModuleDictIndex operator.
+      auto graph = m.graph();
```
We should add clarification to why we’re making it a built in op. 
I think it’s bc: there’s no good way of constructing a dict representing the ModuleDict and caching it across indexes, bc any call to setAttr on the module would invalidate the dict, and we want to preserve the O(1) complexity

eellison commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -822,6 +822,15 @@ struct PythonPrintImpl {
+        const auto dict = node->inputs().at(0);
+        const auto key = node->inputs().at(1);
+        const auto out = node->outputs().at(0);
+        assignValuesToTheirUniqueNames(out);
```
Do we need to call assignValuesToTheirUniqueNames here ?

SplitInfinity replied:
Yes, there's no name for it yet.

eellison commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -653,6 +655,15 @@ RegisterOperators reg(
          aliasAnalysisSpecialCase()),
+     OperatorGenerator(
+         TORCH_SELECTIVE_SCHEMA(
+             "prim::ModuleDictIndex(Any self, str ind) -> Any"),
```
Could you add a comment to this to the extent that this op is generated in the compiler, and self and the output are expected to be set to a ModuleType/ModuleTypeInterface ? I think if you look up TupleIndex there is a similar comment to that extent

eellison reviewed (approved):
LGTM! Sorry there was a little churn but I think we ended up at the right spot. 

Could you add a test in freezing that this errors with a reasonable message ? It should fail when it sees an interface call but I just want to make sure.

Another thing you could add is a peephole optimization that checks if this is a constant key and replaces the node with GetAttr. I suspect that’s not too common, though. Maybe better done as a follow up

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the left-hand side of the assignment has an interface type
annotation and if the dictionary contains only modules that implement
that interface, e.g. `module: ModuleInterface = self.module_dict[key]`.

The approach taken by this commit is that the annotation is passed
into `SugaredValue::getitem` as a type hint whose presence is used to
emit a special `prim::ModuleDictIndex` operator and set the output type
of the node appropriately.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the left-hand side of the assignment has an interface type
annotation and if the dictionary contains only modules that implement
that interface, e.g. `module: ModuleInterface = self.module_dict[key]`.

The approach taken by this commit is that the annotation is passed
into `SugaredValue::getitem` as a type hint whose presence is used to
emit a special `prim::ModuleDictIndex` operator and set the output type
of the node appropriately.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]

SplitInfinity commented:
I tried adding this freezing test and no exception was raised 👀 :
```
    def test_freeze_module_with_non_static_module_dict_index(self):
        """
        Test that a Module contained a non-static ModuleDict index
        cannot be frozen.
        """
        @torch.jit.interface
        class ModuleInterface(torch.nn.Module):
            def forward(self, inp: Any) -> Any:
                pass

        class ImplementsInterface(torch.nn.Module):
            def forward(self, inp: Any) -> Any:
                if isinstance(inp, torch.Tensor):
                    return torch.max(inp, dim=0)

                return inp

        # Test annotation of submodule.
        class Mod(torch.nn.Module):
            def __init__(self):
                super().__init__()
                self.d = torch.nn.ModuleDict({"module": ImplementsInterface()})

            def forward(self, x: torch.Tensor, key: str) -> Any:
                value: ModuleInterface = self.d[key]
                return value.forward(x)

        m = torch.jit.script(Mod())
        m.eval()
        mf = torch._C._freeze_module(m._c)
        with self.assertRaisesRegex(RuntimeError, "Interface"):
            m.forward(torch.randn(4), "module")
```

eellison commented:
Okay, can you just edit https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/freeze_module.cpp#L215 a line here to check for prim::ModuleDictIndex and error saying that freezing models that contain this op is unsupported ?

Commit:
Update on "[JIT] Enable ModuleDict non-literal indexing"


**Summary**
This commit enables indexing into `ModuleDict` using a non-literal
index if the left-hand side of the assignment has an interface type
annotation and if the dictionary contains only modules that implement
that interface, e.g. `module: ModuleInterface = self.module_dict[key]`.

The approach taken by this commit is that the annotation is passed
into `SugaredValue::getitem` as a type hint whose presence is used to
emit a special `prim::ModuleDictIndex` operator and set the output type
of the node appropriately.

**Test Plan**
This commit adds unit tests to `TestModuleContainers` to test this
feature (`test_typed_module_dict`).

Differential Revision: [D24070606](https://our.internmc.facebook.com/intern/diff/D24070606)

[ghstack-poisoned]
